lib/propolis/src/lifecycle.rs

```diff
@@ -95,3 +97,96 @@ pub trait Lifecycle: Send + Sync + 'static {
         Migrator::Empty
     }
 }
+
+/// Indicator for tracking [Lifecycle] states.
+///
+/// As implementors of the [Lifecycle] trait are driven through various state
+/// changes when called through its functions, the underlying resource may want
+/// to keep track of which state it is in.  Device emulation may wish to
+/// postpone certain setup tasks when undergoing state import during a
+/// migration, for example.
+///
+/// The [Indicator] is meant as a convenience tool for such implementors to
+/// simply track the current state by calling the appropriate methods
+/// corresponding to their [Lifecycle] counterparts.
+#[derive(Default)]
+pub struct Indicator(AtomicUsize);
+impl Indicator {
+    pub const fn new() -> Self {
+        Self(AtomicUsize::new(IndicatedState::Init as usize))
+    }
+    /// Indicate that holder is has started
+    ///
+    /// To be called as part of the implementor's [Lifecycle::start()] method
+    pub fn start(&self) {
+        self.state_change(IndicatedState::Run);
+    }
+    /// Indicate that holder is has paused
+    ///
+    /// To be called as when [Lifecycle::pause()] has been called on the
+    /// implementor _and_ that pause has completed (if there is async work
+    /// required for such a state).
+    pub fn pause(&self) {
+        self.state_change(IndicatedState::Pause);
+    }
+    /// Indicate that holder is has resumed
+    ///
+    /// To be called as part of the implementor's [Lifecycle::resume()] method
+    pub fn resume(&self) {
+        self.state_change(IndicatedState::Run);
+    }
+    /// Indicate that holder is has halted
+    ///
+    /// To be called as part of the implementor's [Lifecycle::halt()] method
+    pub fn halt(&self) {
+        self.state_change(IndicatedState::Halt);
+    }
+    /// Get the currently indicated state for the holder
+    pub fn state(&self) -> IndicatedState {
+        IndicatedState::from_usize(self.0.load(Ordering::SeqCst))
+    }
+
+    fn state_change(&self, new: IndicatedState) {
+        let old = self.0.swap(new as usize, Ordering::SeqCst);
+        IndicatedState::assert_valid_transition(
+            IndicatedState::from_usize(old),
+            new,
+        );
+    }
+}
+
+/// Represents the current state held in [Indicator]
+#[derive(Copy, Clone, Eq, PartialEq, strum::FromRepr, Debug)]
+#[repr(usize)]
+pub enum IndicatedState {
+    /// Lifecycle entity has been created, but has not been
+    /// (started)[Lifecycle::start()] yet.
+    Init,
+    /// Lifecycle entity is running
+    Run,
+    /// Lifecycle entity is paused
+    Pause,
+    /// Lifecycle entity has halted
+    Halt,
+}
+impl IndicatedState {
+    const fn valid_transition(old: Self, new: Self) -> bool {
+        use IndicatedState::*;
+        match (old, new) {
+            (Init, Run) | (Init, Halt) => true,
+            (Run, Pause) => true,
+            (Pause, Run) | (Pause, Halt) => true,
+            _ => false,
+        }
+    }
```

Comment:
thinking out loud: this describes the valid state transitions for any `Lifecycle` implementer even if they don't use an `Indicator`. it's an explicit form of the doc comments on `trait Lifecycle`, and feels like we'd get useful error checking if every `Lifecycle` got a free `Indicator` riding along to tell us (panicfully) if a Propolis user gets the state machine wrong. i'm not sure how to spell this in a nice Rust-y way yet, but i agree with the value of `Indicator` and i'll make a followup issue for bringing the implict and explicit state machines together...

(this brought to you by me wondering what should happen when you `reset()`, looking at `Lifecycle`, thinking `Pause -> Run` is prooobably right, then following the transitions in `propolis-standalone` to be more sure.)